File: python/examples/function-pipe_test.py
"""function-pipe: toolz.pipe の Contract を検証する。"""

import asyncio
from collections.abc import Callable
from functools import partial

import pytest
from toolz import compose_left, pipe


def double(n: int) -> int:
    return n * 2


def to_label(n: int) -> str:
    return f"total: {n}"


def test_applies_functions_left_to_right() -> None:
    """data を最初の関数に渡し、以降は直前の戻り値を左から順に渡す。"""
    assert pipe(3, double, to_label) == "total: 6"
    assert pipe(3, to_label, str.upper) == "TOTAL: 3"


def test_executes_synchronously_in_order() -> None:
    """各関数は呼び出し時に同期的に順番どおり実行される。"""
    order: list[str] = []

    def step(name: str) -> Callable[[int], int]:
        def run(n: int) -> int:
            order.append(name)
            return n + 1

        return run

    assert pipe(0, step("a"), step("b"), step("c")) == 3
    assert order == ["a", "b", "c"]


def test_no_functions_returns_data_unchanged() -> None:
    """関数を 0 個で呼ぶと data をそのまま返す。"""
    data = [1, 2]
    assert pipe(data) is data


def test_function_with_two_required_arguments_raises() -> None:
    """各関数は引数 1 つで呼ばれるので、必須引数が 2 つの関数を渡すと TypeError になる。"""

    def add(a: int, b: int) -> int:
        return a + b

    with pytest.raises(TypeError):
        pipe(1, add)


def test_async_function_is_not_awaited() -> None:
    """非同期関数は await されず、コルーチンオブジェクトが次の関数に渡される。"""
    received: list[object] = []

    async def fetch(n: int) -> int:
        return n

    def capture(value: object) -> object:
        received.append(value)
        return value

    result = pipe(1, fetch, capture)
    assert asyncio.iscoroutine(result)
    assert received == [result]
    result.close()


def test_functions_are_not_modified() -> None:
    """渡した関数は変更されず、pipe の後も単独で同じように呼べる。"""
    before = double.__code__
    pipe(1, double, to_label)
    assert double.__code__ is before
    assert double(4) == 8


def test_compose_left_alternative_builds_a_function() -> None:
    """Alternatives: compose_left は値を渡さずに合成した関数を作り、pipe と同じ結果になる。"""
    summarize = compose_left(double, to_label)
    assert summarize(3) == pipe(3, double, to_label)
    assert compose_left()(5) == 5


def test_partial_for_multi_argument_step() -> None:
    """Alternatives: 複数引数の関数は partial で残り 1 引数にしてから渡す。"""
    assert pipe([1, 2, 3], partial(map, double), list) == [2, 4, 6]

File: typescript/examples/function-pipe.test.ts
// カード function-pipe の Contract を検証するテスト
import { flow, flowRight } from 'es-toolkit';
import { describe, expect, it } from 'vitest';

describe('function-pipe: flow', () => {
  it('最初の関数は全引数を受け取り、以降は直前の戻り値を左から順に受け取る', () => {
    const order: string[] = [];
    const add = (a: number, b: number) => {
      order.push('add');
      return a + b;
    };
    const double = (n: number) => {
      order.push('double');
      return n * 2;
    };
    const toLabel = (n: number) => `total: ${n}`;
    expect(flow(add, double, toLabel)(1, 2)).toBe('total: 6');
    expect(order).toEqual(['add', 'double']);
  });

  it('関数 0 個なら第 1 引数をそのまま返す', () => {
    expect(flow()(7, 8)).toBe(7);
  });

  it('5 個までは型が推論される', () => {
    const f = flow(
      (s: string) => s.length,
      (n) => n + 1,
      (n) => n * 2,
      (n) => String(n),
      (s) => s.padStart(4, '0'),
    );
    const result: string = f('abc');
    expect(result).toBe('0008');
  });

  it('this をすべての関数に渡す', () => {
    const obj = {
      base: 10,
      run: flow(
        function (this: { base: number }, n: number) {
          return this.base + n;
        },
        function (this: { base: number }, n: number) {
          return this.base * n;
        },
      ),
    };
    expect(obj.run(1)).toBe(110);
  });

  it('非同期関数を混ぜても await せず Promise がそのまま次に渡る', async () => {
    const received: unknown[] = [];
    const f = flow(
      async (n: number) => n + 1,
      (p) => {
        received.push(p);
        return p;
      },
    );
    const result = f(1);
    expect(received[0]).toBeInstanceOf(Promise);
    expect(await result).toBe(2);
  });

  it('flowRight は右から左に合成する', () => {
    const toLabel = (n: number) => `total: ${n}`;
    const double = (n: number) => n * 2;
    const add = (a: number, b: number) => a + b;
    expect(flowRight(toLabel, double, add)(1, 2)).toBe('total: 6');
  });

  it('渡した関数を変更せず、合成結果は新しい関数', () => {
    const inc = (n: number) => n + 1;
    const f = flow(inc);
    expect(f).not.toBe(inc);
    expect(inc(1)).toBe(2);
  });
});
